File: packages/i18nify-js/src/modules/dateTime/__tests__/parseFlexibleDate.test.ts
import parseFlexibleDate from '../parseFlexibleDate';

describe('parseFlexibleDate', () => {
  it('accepts a Date object and returns structured fields', () => {
    const date = new Date(2024, 0, 15, 10, 30, 45);
    const result = parseFlexibleDate(date);

    expect(result.year).toBe(2024);
    expect(result.month).toBe(1);
    expect(result.day).toBe(15);
    expect(result.hour).toBe(10);
    expect(result.minute).toBe(30);
    expect(result.second).toBe(45);
  });

  it('accepts a Unix timestamp in milliseconds', () => {
    const timestamp = new Date(2021, 0, 1).getTime();
    const result = parseFlexibleDate(timestamp);

    expect(result.year).toBe(2021);
    expect(result.month).toBe(1);
    expect(result.day).toBe(1);
    expect(result.timestamp).toBe(timestamp);
  });

  it('parses supported date strings', () => {
    expect(parseFlexibleDate('2024-01-15').day).toBe(15);
    expect(parseFlexibleDate('15/01/2024').month).toBe(1);
  });

  it('throws for empty or invalid input', () => {
    expect(() => parseFlexibleDate('' as never)).toThrow(
      'parseFlexibleDate: invalid input',
    );
    expect(() => parseFlexibleDate('not-a-date')).toThrow();
  });
});

File: packages/i18nify-go/modules/datetime/parse_flexible_date_test.go
package datetime

import (
	"testing"
	"time"
)

func TestParseFlexibleDate(t *testing.T) {
	previousLocal := time.Local
	time.Local = time.UTC
	defer func() {
		time.Local = previousLocal
	}()

	t.Run("parses time.Time input", func(t *testing.T) {
		input := time.Date(2024, time.January, 15, 10, 30, 45, 0, time.UTC)
		result, err := ParseFlexibleDate(input)
		if err != nil {
			t.Fatalf("expected no error, got %v", err)
		}

		if result.Year != 2024 || result.Month != 1 || result.Day != 15 {
			t.Fatalf("unexpected parsed date: %+v", result)
		}
	})

	t.Run("parses millisecond timestamp input", func(t *testing.T) {
		result, err := ParseFlexibleDate(int64(1609459200000))
		if err != nil {
			t.Fatalf("expected no error, got %v", err)
		}

		if result.Year != 2021 || result.Month != 1 || result.Day != 1 {
			t.Fatalf("unexpected parsed date: %+v", result)
		}
	})

	t.Run("parses supported string layouts", func(t *testing.T) {
		result, err := ParseFlexibleDate("2024-01-15")
		if err != nil {
			t.Fatalf("expected no error, got %v", err)
		}

		if result.Day != 15 {
			t.Fatalf("unexpected parsed date: %+v", result)
		}
	})

	t.Run("rejects invalid input", func(t *testing.T) {
		if _, err := ParseFlexibleDate(""); err == nil {
			t.Fatal("expected error for empty input")
		}

		if _, err := ParseFlexibleDate("not-a-date"); err == nil {
			t.Fatal("expected error for invalid date string")
		}
	})
}
